Mathlib/Data/Greedoid/Predefs.lean

```diff
@@ -0,0 +1,193 @@
+/-
+Copyright (c) 2024 Jihoon Hyun. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jihoon Hyun
+-/
+import Mathlib.Data.Finset.Basic
+import Mathlib.Data.Finset.Card
+
+/-!
+# Exchange Property and Accessible Property of Greedoid
+Greedoid is a set system, i.e., a family of sets, over a finite ground set, which contains
+an empty set and satisfies both exchange and accessible properties.
+If a set system `S` satisfies the exchange property, then there is some element `x ∈ s₂ \ s₁`
+which `s₁ ∪ {x} ∈ S`, for every set `s₁, s₂ ∈ S` satisfying `s₁.card < s₂.card`.
+If a set system `S` satisfies the accessible property, then there is some element `x ∈ s`
+which `s \ {x} ∈ S` for every nonempty set `s ∈ S`.
+These two properties are defined in this file as `ExchangeProperty` and `AccessibleProperty`.
+
+While it is sufficient to define a greedoid using only the definitions of the properties, it turned
+out to be useful to have some theorems equipped with them.
+Therefore this file defines exchange property and accessible property, and hands out some basic
+theorems which come along with it, prior to defining a greedoid.
+-/
+
+namespace Greedoid
+
+open Nat Finset
+
+/-- The exchange property of greedoid.
+    Note that the exchange property also hold for matroids. -/
+def ExchangeProperty {α : Type _} [DecidableEq α] (Sys : Finset (Finset α)) :=
+  {s₁ : Finset α} → s₁ ∈ Sys →
+  {s₂ : Finset α} → s₂ ∈ Sys →
+  s₂.card < s₁.card →
+  ∃ x ∈ s₁ \ s₂, insert x s₂ ∈ Sys
+
+instance {α : Type _} [DecidableEq α] :
+    @DecidablePred (Finset (Finset α)) ExchangeProperty :=
+  fun Sys =>
+    if h : ∃ s₁ ∈ Sys, ∃ s₂ ∈ Sys, s₂.card < s₁.card ∧ ∀ x ∈ s₁ \ s₂, insert x s₂ ∉ Sys
+    then isFalse (fun h' => by
+      let ⟨s₁, hs₁, s₂, hs₂, hs₃, hs₄⟩ := h
+      have ⟨_, ha₁, ha₂⟩ := h' hs₁ hs₂ hs₃
+      exact hs₄ _ ha₁ ha₂)
+    else isTrue (by
+      simp at h
+      intro _ hs₁ _ hs₂ hs
+      have ⟨a, ha⟩ := h _ hs₁ _ hs₂ hs
+      exists a; simp only [mem_sdiff, ha, not_false_eq_true, and_self])
```

Comment:
Unless you're trying to define a computationally efficient instance, this works:
```suggestion
  fun Sys => by unfold ExchangeProperty; infer_instance
```
If you want to define a computationally efficient instance, then please add a docstring saying so.